### DISCUSSION TITLE
Compatibility issue(?)

### DISCUSSION THREAD

**Discussion by EnderWarik** [Quillraven/Fleks#112] [Q&A] [answered]
I am new to this field (game development) and not very experienced in working with java.  I do everything following the example of the author, Mystic wood, I watch videos on his YouTube channel for the old version of fleks, and in parallel I try to transfer to the new one myself (I know that there is code on the git already from the new versions). in general, I came across an incompatibility problem, As indicated in the documentation of the 2x version, I have two ways to do family filtering:
```
// 1) via the 'compareEntity' function
class RenderSystem : IteratingSystem(
    family = family { all(Position, Render) },
    comparator = compareEntity { entA, entB -> entA[Position].y.compareTo(entB[Position].y) }
) {
    override fun onTickEntity(entity: Entity) {
        // render entities: entities are sorted by their y-coordinate
    }
}

// 2) via the 'compareEntityBy' function which requires the component to implement the Comparable interface
data class Position(
    var x: Float,
    var y: Float,
) : Component<Position>, Comparable<Position> {
    override fun type() = Position

    override fun compareTo(other: Position): Int {
        return y.compareTo(other.y)
    }

    companion object : ComponentType<Position>()
}

class RenderSystem : IteratingSystem(
    family = family { all(Position, Render) },
    comparator = compareEntityBy(Position)
) {
    override fun onTickEntity(entity: Entity) {
        // render entities: entities are sorted by their y-coordinate
    }
}
```
But both of these cases lead to a problem: `Cannot inline bytecode built with JVM target 11 into bytecode that is being built with JVM target 1.8. Please specify proper '-jvm-target' option`
The project was created as in the video, on java 8, checked the project structure, settings->kotlin compiler, version 1.8 is indicated everywhere
Changing these parameters in java 11 does not help. What could be the problem?
Created via gdx-liftoff desktop version
gradle.properties:
```
org.gradle.daemon=true
org.gradle.jvmargs=-Xms512M -Xmx1G
org.gradle.configureondemand=false
kotlinVersion=1.8.21
aiVersion=1.8.2
box2dlightsVersion=1.5
ktxVersion=1.12.0-rc1
gdxVersion=1.12.0
fleksVersion=Fleks:2.4
```
all code: https://github.com/EnderWarik/TheEnderGame

**Comment by Quillraven** [comment]
Hi
Sorry i am on my phone right now. Please have a look at one of the last commits in MysticGarden or MysticWoods 2.0 branch.

You need to update the gradle files properly to set bytecode level to 11 instead of 8. I cannot remember the lines and will add them later and also add this info to the wiki.

But you will find them in the commits in the projects mentioned above. The commits that update Fleks to 2.4..

Manually changing it in the IDE won't work. At least it did not work for me.

**Comment by EnderWarik** [comment]
Thanks for the reply.
I needed to take the following steps to make it work 
1. change the gradle project:
```
configure(subprojects) {
  apply plugin: 'java-library'
  apply plugin: 'kotlin'
  sourceCompatibility = 11
  compileJava {
    options.incremental = true
  }
  compileKotlin {
    kotlinOptions.jvmTarget = "11"
  }
  compileTestKotlin {
    kotlinOptions.jvmTarget = "11"
  }
}
```
After that, the error remains underlined, but the application starts. To remove it, it came to manually remove the dependency from gradle core, build, return dp and build again. Gradle clean -> build doesn't help

Thank you, the issue is resolved


UP: At the same time, in file->project structure -> sdk should have the 11th version

**Reply by Quillraven** [reply, marked as answer]
Great that you figured it out yourself. I also tried to create a new project now from scratch using the latest version of [gdx-liftoff](https://github.com/tommyettinger/gdx-liftoff/releases).

The things I changed after the project was created:
- add `fleksVersion` of 2.4 to `gradle.properties`
- add `api "io.github.quillraven.fleks:Fleks:$fleksVersion"` to build.gradle of the `core` project
- Reload gradle projects via the "refresh" button in the gradle tab of IntelliJ
- Pick correct SDK under `File -> Project Structure -> Project Settings -> Project -> SDK`. It must be at least Java 11 SDK

That worked for me since gdx-liftoff is now already creating projects with Java 11. I also had a look at the files and following parts are important:
- root `build.gradle`: sourceCompatibility and jvmToolchain. Not sure if jvmToolchain works with gradle below version 8. The alternative is `compileKotlin` and `compileTestKotlin` like you had it above:
  ```
  configure(subprojects) {
    apply plugin: 'java-library'
    apply plugin: 'kotlin'
    sourceCompatibility = 11
    compileJava {
      options.incremental = true
    }
  
    kotlin {
      jvmToolchain(11)
    }
  }
  ```
- `sourceCompatibility = 11` in `lwjgl3` project
